Binning

philippemiron opened the pull request.

philippemiron commented:
- [] add back the digitize
- [] extending to more statistics than just the current mean ([See, `twodstats.m`](https://github.com/jonathanlilly/jLab/blob/master/jStats/twodstats.m)
- [] better function and parameter names
- [] automatic handling of datetime in coordinates/variables
- [] add tests

philippemiron commented:
@miniufo, true, those parameters could be combined. I was trying to follow the numpy [`histogramdd`](https://numpy.org/doc/2.0/reference/generated/numpy.histogramdd.html) API for consistency. I think it's simpler to pass integer(s) to the `bins` when you want to see the full range of the data.

selipot commented:
Not the behavior I expected. Not sure what is going on for me. Why only one coordinate and one data variable for the output?
```
from clouddrift.binning import histogram

coords = [ds.lon, ds.lat]
variables = [ds.ve, ds.vn]

ds_binned = histogram(
    coords=coords,
    data=variables,
    zeros_to_nan=True,
)

ds_binned
```
<img width="883" alt="Screenshot 2025-06-23 at 4 13 02 PM" src="https://github.com/user-attachments/assets/fd176d9d-8c3b-4988-985c-b547ae85cddd" />

philippemiron commented:
Ok, I see the bug.

Try:

```
coords = [ds.lon.values, ds.lat.values]
variables = [ds.ve.values, ds.vn.values]
```

philippemiron commented:
9.4s on my Mac Mini (M4 Pro)

![image](https://github.com/user-attachments/assets/fc360f08-2323-4296-91d0-3fb6a171dd31)

selipot commented:
@philippemiron sorry, instead of pushing to the PR I could have made code change suggestions?

philippemiron commented:
> @philippemiron sorry, instead of pushing to the PR I could have made code change suggestions?

No worries!

philippemiron commented on clouddrift/binning.py:
```diff
@@ -0,0 +1,154 @@
+        Outer bin limits for each dimension\
+    dim_names : list of str, optional
+        Names for the dimensions of the output xr.Dataset
+        If None, default names are "dim_0_bin", "dim_1_bin", etc.
```
I'm not sure I understand this.

selipot replied:
`coords=[x,y,z] -> dim_names=[x_bin, y_bin,z_bin, ...]`
`data=[a,b,c] -> output_names=[a_mean,b_mean,c_mean]`

philippemiron replied:
I don't think that's possible. Inside the function, we only have access to the data and not the original variable names.

philippemiron commented on clouddrift/binning.py:
```diff
@@ -0,0 +1,154 @@
+DEFAULT_BINS_NUMBER = 10
+
+
+def histogram(
```
You are correct.. but I think we should find another name because next up is to add the possibility of calculating different statistics on the dataset, maybe we get inspiration from `SciPy`, https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binned_statistic.html#scipy.stats.binned_statistic.

`binned_statistic()` ?

philippemiron commented on clouddrift/binning.py:
```diff
@@ -0,0 +1,154 @@
+
+    Returns
+    -------
+    xr.Dataset
```
Once we implement different functions, maybe we can pass the "statistics" as a list, so something like : `statistics = "count", "mean"` would return both.

selipot replied:
Oh I see, I missed that. That is not explained in the docstring. So one would have to call the function twice? I think the bin count should be added to the dataset when the means are calculated.

philippemiron replied:
Yes, for now I'm thinking we should be able to pass any `Callable` that does:
```
def func(data: array-like): -> float
```

philippemiron commented:
I added the following:

```
coords = [ds.lon, ds.lat]
variables = [ds.ve, ds.vn]
bins = [360, 180]

ds_binned = binned_statistics(
    coords=coords,
    bins=bins,
)
```

```
> print(ds_binned)

<xarray.Dataset> Size: 523kB
Dimensions:       (dim_0_bin: 360, dim_1_bin: 180)
Coordinates:
  * dim_0_bin     (dim_0_bin) float64 3kB -179.5 -178.5 -177.5 ... 178.5 179.5
  * dim_1_bin     (dim_1_bin) float64 1kB -77.84 -76.9 -75.97 ... 88.58 89.52
Data variables:
    binned_count  (dim_0_bin, dim_1_bin) int64 518kB 32 117 59 72 ... 61 129 4 0
```

where I changed the default name to `binned_count`.

Then, if you pass `data`.

```
ds_binned = binned_statistics(
    coords=coords,
    data=variables,
    bins=bins,
    bins_range=[(-180, 180), (-90, 90)],
    dim_names=["lon_bin", "lat_bin"],
    output_names=["binned_mean_ve", "binned_mean_vn"],
    zeros_to_nan=True,
)
```

```
> print(ds_binned)

<xarray.Dataset> Size: 2MB
Dimensions:               (lon_bin: 360, lat_bin: 180)
Coordinates:
  * lon_bin               (lon_bin) float64 3kB -179.5 -178.5 ... 178.5 179.5
  * lat_bin               (lat_bin) float64 1kB -89.5 -88.5 -87.5 ... 88.5 89.5
Data variables:
    binned_mean_ve_count  (lon_bin, lat_bin) int64 518kB 0 0 0 0 ... 120 13 0
    binned_mean_ve        (lon_bin, lat_bin) float64 518kB nan nan ... nan
    binned_mean_vn_count  (lon_bin, lat_bin) int64 518kB 0 0 0 0 ... 120 13 0
    binned_mean_vn        (lon_bin, lat_bin) float64 518kB nan nan ... nan
```
I'm have now included the `_count` for each variable. Note that those can differ if a data point is `nan` along the path.

selipot commented on clouddrift/binning.py:
```diff
@@ -0,0 +1,177 @@
+    zeros_to_nan: bool = False,
+) -> xr.Dataset:
+    """
+    Compute N-dimensional histogram binning and calculate variable means in each bin.
```
Perform N-dimensional binning and compute mean of values in each bin. The result is returned as an Xarray Dataset.

selipot commented on clouddrift/binning.py:
```diff
@@ -0,0 +1,177 @@
+    Parameters
+    ----------
+    coords : array-like or list of array-like
+        Arrays of the coordinates of the Lagrangian data to be binned. For 1D data, a single array can be provided.
```
Array(s) of Lagrangian data coordinates to be binned. For 1D, provide a single array. For N-dimensions, provide a list of N arrays, each giving coordinates along one dimension.

selipot commented on clouddrift/binning.py:
```diff
@@ -0,0 +1,177 @@
+        For multiple dimensions, the first array represents each coordinates along the first dimension,
+        the second array represents each coordinates along the second dimension, and so on.
+    data : array-like or list of array-like
+        Data associated at the Lagrangian coordinates described in coords. Multiple variables can be provided as a list.
```
Data values associated with the Lagrangian coordinates in coords. Can be a single array or a list of arrays for multiple variables

selipot commented on clouddrift/binning.py:
```diff
@@ -0,0 +1,177 @@
+    data : array-like or list of array-like
+        Data associated at the Lagrangian coordinates described in coords. Multiple variables can be provided as a list.
+    bins : int or lists, optional
+        Number of bins per dimension (int) or bin edges per dimension (list). Default is 10.
```
“Number of bins or bin edges per dimension. It can be:
	•	An int: same number of bins for all dimensions (default: 10),
	•	A list of ints or arrays: one per dimension, specifying either bin count or bin edges,
	•	None: defaults to 10 bins per dimension.”

selipot commented on clouddrift/binning.py:
```diff
@@ -0,0 +1,177 @@
+        Each element can be an integer or an array of bin edges.
+        If None, defaults to 10 bins per dimension.
+    bins_range : list of tuples, optional
+        Outer bin limits for each dimension\
```
typo: Outer bin limits for each dimension.

selipot reviewed (commented):
I made some suggestions for the docstring. I notice the function returns a count for each variable. In which case do the counts differ? If a variable has more nan values than the other as an example?

selipot commented:
Looks good to me to merge! @KevinShuman please review.

KevinShuman reviewed (approved):
All looks good!
